File: Graphs/LC_323_Number_of_Connected_Components_in_an_Undirected_Graph.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.18

class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        vector<int> root;
        for (int i = 0; i < n; i++){
            root.push_back(i);
        }
        for(vector<int> edge: edges){
            if (doUnion(edge[0], edge[1], root)){
                n -= 1;
            }
        }
        return n;
    }
private:
    int find(int x, vector<int>& root){
        if (x != root[x]){
            root[x] = find(root[x], root);
        }
        return root[x];
    }

    bool doUnion(int x, int y, vector<int>& root){
        int rx = find(x, root), ry = find(y, root);
        if (rx != ry){
            root[ry] = rx;
            return true;
        }
        return false;
    }
};

File: Graphs/LC_323_Number_of_Connected_Components_in_an_Undirected_Graph.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.18

class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        root = [i for i in range(n)]
        total = n

        def find(x):
            if root[x] != x:
                root[x] = find(root[x])
            return root[x]
        
        def union(x, y):
            rx, ry = find(x), find(y)
            nonlocal total
            if rx != ry:
                root[ry] = rx
                total -= 1

        for v1, v2 in edges:
            union(v1, v2)
        return total